### DISCUSSION TITLE
MediaElement full-screen orientation

### DISCUSSION THREAD

**Discussion by armanimichael** [CommunityToolkit/Maui#1945] [New Feature Discussions]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Did you read the "Reporting a bug" section on Contributing file?

- [X] I have read the "Reporting a bug" section on Contributing file: https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#reporting-a-bug

### Current Behavior

If Android is locked in Portrait mode, MediaElement full-screen won't rotate the video horizontally.
[See example video](https://github.com/CommunityToolkit/Maui/assets/25107976/b79f5ce7-9314-45c0-8ae2-c0b082c9c117)

### Expected Behavior

Regardless of orientation lock, a video should follow the screen orientation when full-screen.

### Steps To Reproduce

With `ScreenOrientation = ScreenOrientation.Portrait` 

1. Open video in fullscreen
2. Rotate the device into landscape mode

### Link to public reproduction project repository

https://github.com/armanimichael/MediaElementFullScreen

### Environment

```markdown
- .NET MAUI CommunityToolkit.Maui.MediaElement: 3.1.0
- .NET MAUI CommunityToolkit: 8.0.0
- OS: Android 14
- .NET MAUI: 8.0.14
```

**Comment by ne0rrmatrix** [comment]
That is default behavior inherited from android. I opened 3 different media players on android. Youtube, twitch, and Prime Video. The all have the exact same behavior that you describe. I don't see any reason to change that. If the device is rotation locked it should not rotate the video when you change orientation. That is expected and default behavior.

If we change this it would require overriding default behavior and I don't think that is a good idea.

**Comment by armanimichael** [comment]
Thanks for the response @ne0rrmatrix.

I think there's some confusion here, I don't see any issue with this if the device rotation is being disabled **by the user.**

What looks wrong to me here is preventing the video from adapting to the rotation when **I (as a developer) lock the Application to portrait mode** (for instance because landscape is not supported for my use-case / UI).

I checked Youtube and Prime Video from your examples, and those applications do support both portrait and landscape mode.

On IOS for instance, `MediaElement` handles this scenario perfectly (if the application is locked to portrait mode, the video still adapts to landscape)

**Comment by ne0rrmatrix** [comment]
Yes I think I am confused. Are you saying that you set orientation lock and the android implementation is staying locked? If you lock orientation in Maui you want the video player specifically to ignore that? I'm not sure exactly what it is you are expecting?

**Comment by ne0rrmatrix** [comment]
I just double-checked and when screen rotation is locked the video does not rotate in Android in a video player by default.  I loaded the sample app from repo in portrait mode in both android and iOS. Both exhibit the same behavior. If I lock rotation from the menu it stays in that rotation for the whole time screen rotation is locked.

edit for clarity: I opened the sample app from the Maui Community toolkit. Not the provided sample. Sorry about the any possible confusion. I am now going to do further testing with the provided sample to see the behavior.

**Comment by armanimichael** [comment]
There are two ways to lock the application in portrait mode, by two different actors:

- Developers can specify that on `MainActivity` in the Activity attribute by using `ScreenOrientation = ScreenOrientation.Portrait`
- Users can use the system settings toggle to enable/disable changing the orientation
    - On top of this - when locked - the user can force the orientation if they are using the bottom bar with buttons: an icon appears on the right and they can force either landscape or portrait mode  (see attached image)

 
Now, if you try YouTube app on an Android phone. there's no way to use it in landscape mode besides entering video fullscreen mode while watching a landscape video (if the video has portrait proportions, the full screen will remain in portrait mode).
In all other use cases the YouTube app is locked in portrait mode.

Let's say I want to create a YouTube clone.
First thing I would do is go into `MainActivity` and - as a developer - specify `ScreenOrientation = ScreenOrientation.Portrait`.

That said, I want my users to be able to tap on the fullscreen button in the video player, and:

- Nice to have: that should change the screen orientation forcefully depending if the video is a landscape one
- Good enough: that should ignore the system setting (set by the user) and detect screen orientation change and display the video accordingly

![Screenshot_20240404-165555](https://github.com/CommunityToolkit/Maui/assets/25107976/ecb78050-7aea-4f15-9b49-534ad64ed83c)

**Comment by ne0rrmatrix** [comment]
Ok. That wouuld be easy to implement as I already have an idea how to do that. Sorry about the confusion. I would be happy to submit a proposal to add the ability to do force rotation like that as an option for developers. I will talk to a team member this morning.

**Comment by armanimichael** [comment]
Awesome, thank you @ne0rrmatrix!

**Comment by ne0rrmatrix** [comment]
I just checked both my phone and android tablet. The behavior of the device is  not as you describe. If I turn on auto rotation in settings it will auto rotate based on orientation in YouTube. This is default behavior of most video apps on Samsung devices. This works for both full screen and not full screen. 

I am going to start working on suggested change to allow developers a choice to lock this behavior based on choice. It will be a developer option and not a default behavior. 

Edit: I will create a discussion to have the team and community discuss it. I have not yet started working on it.

**Comment by Neotrickster** [comment]
I have the same case, my app is Portrait for everything but 1 page (the one with MediaElement) so I just add some android specific code in OnAppearing and OnDisappearing so change only when enter the page and leave the one who have before changed when leave the page.

I just see @ne0rrmatrix have a OnFullscreenButtonClick method mentioned but I don´t know atm how to access it, but I´m guessing that method will be a better place to put the code I´m refereing.

Hopefully this will not be needed at all when a better solution founded but maybe works for some people in the meantime.

```
object SOToRestore {  get; set; }

protected override void OnAppearing()
{
    base.OnAppearing();

#if ANDROID
SOToRestore = Microsoft.Maui.ApplicationModel.Platform.CurrentActivity.RequestedOrientation;
Microsoft.Maui.ApplicationModel.Platform.CurrentActivity.RequestedOrientation = Android.Content.PM.ScreenOrientation.Unspecified;
#endif
}

protected override void OnDisappearing()
{
    base.OnDisappearing();

#if ANDROID
    if (SOToRestore is Android.Content.PM.ScreenOrientation SO) Microsoft.Maui.ApplicationModel.Platform.CurrentActivity.RequestedOrientation = Android.Content.PM.ScreenOrientation.Portrait;
#endif
}
```

**Comment by Neotrickster** [comment]
> I just checked both my phone and android tablet. The behavior of the device is not as you describe. If I turn on auto rotation in settings it will auto rotate based on orientation in YouTube. This is default behavior of most video apps on Samsung devices. This works for both full screen and not full screen.
> 
> I am going to start working on suggested change to allow developers a choice to lock this behavior based on choice. It will be a developer option and not a default behavior.
> 
> Edit: I will create a discussion to have the team and community discuss it. I have not yet started working on it.

Is the discussion setting already ?? I can´t find any.

**Comment by Neotrickster** [comment]
>... 
>     * Nice to have: that should change the screen orientation forcefully depending if the video is a landscape one
> 
>     * Good enough: that should ignore the system setting (set by the user) and detect screen orientation change and display the video accordingly
> ...

I just add in "Nice to have" a way to know if the FullScreen button is clicked or a flag to know if MediaElement is on FullScreen/Normal, atm just have OnSizeAllocated but don´t help that much without the flag/method

**Comment by Neotrickster** [comment]
> I just checked both my phone and android tablet. The behavior of the device is not as you describe. If I turn on auto rotation in settings it will auto rotate based on orientation in YouTube. This is default behavior of most video apps on Samsung devices. This works for both full screen and not full screen.
> 
> I am going to start working on suggested change to allow developers a choice to lock this behavior based on choice. It will be a developer option and not a default behavior.
> 
> Edit: I will create a discussion to have the team and community discuss it. I have not yet started working on it.

Just for reference related: https://github.com/CommunityToolkit/Maui/discussions/1748#discussion-6363780

**Comment by ne0rrmatrix** [comment]
Feel free to create a discussion to start getting the ball rolling @Neotrickster

**Comment by Neotrickster** [comment]
Thanks man

**Comment by ne0rrmatrix** [comment]
@vhugogarcia I think this one can be moved to a discussion?
